lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1107 @@
+  return writeMatMul(node, graph, reporter);
+}
+
+llvm::Error ONNXModelWriter::writeReshape(const Node *node, GraphType &graph,
+                                          ReporterType &) {
+  // Dimensions into the constant tensor
+  const auto *RN = llvm::cast<ReshapeNode>(node);
+
+  auto dims = RN->getDims();
+  Tensor oneDimTensor(ElemKind::Int64ITy, {dims.size()});
+  auto handle = oneDimTensor.getHandle<int64_t>();
+  for (size_t b = 0, e = dims.size(); b < e; ++b) {
+    handle.raw(b) = dims[b];
+  }
+  auto *tensorProto = graph.add_initializer();
+  tensorProto->set_name(node->getName());
+  writeTensor(oneDimTensor, tensorProto);
+
+  return writeAll("Reshape", node, graph);
+}
+
+llvm::Error ONNXModelWriter::writeSoftMax(const Node *node, GraphType &graph,
+                                          ReporterType &) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type("Softmax");
+  outputsToProto(node, proto);
+  // Find input from Reshape node
+  proto->add_input(
+      node->getNthInput(ReshapeNode::InputIdx).getNode()->getName());
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeReplaceNaN(const Node *node, GraphType &graph,
+                                             ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *RNANN = llvm::cast<ReplaceNaNNode>(node);
+
+  float value = RNANN->getValue();
+  if (value != 0.0f) {
+    addValueAttribute(proto, "value", value);
+  }
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeGatherRanges(const Node *node,
+                                               GraphType &graph,
+                                               ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *GRN = llvm::cast<GatherRangesNode>(node);
+
+  addValueAttribute(
+      proto, "maxOutputSize",
+      GRN->getOutput().getNode()->dims(GatherRangesNode::OutputIdx)[0]);
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeSparseToDenseMask(const Node *node,
+                                                    GraphType &graph,
+                                                    ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *STDMN = llvm::cast<SparseToDenseMaskNode>(node);
+
+  addValueAttribute(proto, "mask", STDMN->getMask());
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeLocalResponseNormalization(
+    const Node *node, GraphType &graph, ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type("LRN");
+  outputsToProto(node, proto);
+  // Find input from Transpose node
+  const Node *TN =
+      node->getNthInput(LocalResponseNormalizationNode::InputIdx).getNode();
```

Comment:
Same -- prefer using `llvm::dyn_cast` here and using the Node itself below. I.e. once `writeLocalResponseNormalization()` passes in an actual `LRNNode *` instead of a `Node *`, could simply be `const TransposeNode *TN = llvm::dyn_cast<TransposeNode>(LRN->getInput());`